### PR TITLE
fix(explore): time aggregations group by

### PR DESCRIPTION
**Pull request by skjindal93**

### CODE REVIEW

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/RequestHandler.java [hunk 1]
```diff
@@ -36,6 +36,7 @@
 
 public class RequestHandler implements RequestHandlerWithSorting {
   private static final Logger LOG = LoggerFactory.getLogger(RequestHandler.class);
+
```
Comment (sarthak77) [inline gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/RequestHandler.java:39]: Is this intentional?

Comment (skjindal93) [reply gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/RequestHandler.java:39]: yes, I prefer to keep a gap between static and non static class variables

File: gateway-service-impl/src/test/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+import org.hypertrace.gateway.service.v1.explore.ExploreRequest;
+import org.hypertrace.gateway.service.v1.explore.ExploreResponse;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+
+public class TimeAggregationsWithGroupByRequestHandlerTest {
+  @Mock private AttributeMetadataProvider attributeMetadataProvider;
+  @Mock private RequestHandler normalRequestHandler;
+  @Mock private TimeAggregationsRequestHandler timeAggregationsRequestHandler;
```
Comment (sarthak77) [inline gateway-service-impl/src/test/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandlerTest.java:38]: nit : can remove mock annotation.

**Review by sarthak77 — DISMISSED** [review]
LGTM

File: gateway-service-impl/src/test/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandlerTest.java (hunk 2 above)
Comment (suresh-prakash) [reply gateway-service-impl/src/test/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandlerTest.java:38]: +1

**Review by suresh-prakash — DISMISSED** [review]

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandler.java [hunk 3]
```diff
@@ -1,31 +1,54 @@
...
 
   TimeAggregationsWithGroupByRequestHandler(
       QueryServiceClient queryServiceClient, AttributeMetadataProvider attributeMetadataProvider) {
+    this.attributeMetadataProvider = attributeMetadataProvider;
     this.normalRequestHandler = new RequestHandler(queryServiceClient, attributeMetadataProvider);
     this.timeAggregationsRequestHandler =
         new TimeAggregationsRequestHandler(queryServiceClient, attributeMetadataProvider);
   }
 
+  @VisibleForTesting
```
Comment (aaron-steinfeld) [inline gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandler.java:42]: 😢 (this version is actually better, but I'd rather remove the old constructor and make the actual code use this one too)

File: gateway-service-impl/src/test/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandlerTest.java (hunk 2 above)
Comment (aaron-steinfeld) [reply gateway-service-impl/src/test/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandlerTest.java:38]: opposite nit: remove the setup step and leave the mock annotations (you're missing the class annotation which is why the mocks aren't working)

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandler.java [hunk 4]
```diff
@@ -102,42 +129,143 @@ private ExploreRequest buildTimeAggregationsRequest(
...
+
+    if (!resultKeyToAttributeMetadataMap.containsKey(groupByResultName)) {
+      return Optional.empty();
+    }
+
+    AttributeMetadata groupByAttributeMetadata =
+        resultKeyToAttributeMetadataMap.get(groupByResultName);
+    // RHS value of in clause filter should always be an array to apply IN filter clause on groupBy
+    // expression
+    Value.Builder valueBuilder = Value.newBuilder();
```
Comment (aaron-steinfeld) [inline gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandler.java:173]: can we move this out from being inline in a giant method. Ideally into its own class.

Comment (aaron-steinfeld) [reply gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandler.java:173]: I realize now that this is really the whole point of this method, but the generally point I think remains - it's a ~100 line method with multiple giant switches in it. Must be a way to clean that up.

File: gateway-service-impl/src/test/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandlerTest.java (hunk 2 above)
Comment (sarthak77) [reply gateway-service-impl/src/test/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandlerTest.java:38]: >remove the setup step

I think we need to open and close the mocks then.

Comment (aaron-steinfeld) [reply gateway-service-impl/src/test/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandlerTest.java:38]: Mockito does that automatically if you use 
```
@ExtendWith(MockitoExtension.class)
```
It also makes tests strict and all kinds of other goodies - we really shouldn't be writing new tests without it.

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandler.java [hunk 5]
```diff
@@ -102,42 +128,148 @@ private ExploreRequest buildTimeAggregationsRequest(
...
+        remapAttributeMetadataByResultName(
+            originalRequestContext.getExploreRequest(), attributeMetadataMap);
+
+    if (!resultKeyToAttributeMetadataMap.containsKey(groupByResultName)) {
+      return Optional.empty();
+    }
+
+    AttributeMetadata groupByAttributeMetadata =
+        resultKeyToAttributeMetadataMap.get(groupByResultName);
+    // RHS value of in clause filter should always be an array to apply IN filter clause on groupBy
```
Comment (sarthak77) [inline gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandler.java:170]: This comment is present twice. Ig can be removed from here.

**Review by aaron-steinfeld — APPROVED** [review]

**Pull request merged**